File: packages/ocp-shuffle/src/elgamal.ts
import type { ElGamalCiphertext, GroupElement, Scalar } from "@onchainpoker/ocp-crypto";
import { mulBase, mulPoint, pointAdd } from "@onchainpoker/ocp-crypto";

export function elgamalKeygen(sk: Scalar): { sk: Scalar; pk: GroupElement } {
  return { sk, pk: mulBase(sk) };
}

export function elgamalReencrypt(pk: GroupElement, ct: ElGamalCiphertext, rho: Scalar): ElGamalCiphertext {
  return {
    c1: pointAdd(ct.c1, mulBase(rho)),
    c2: pointAdd(ct.c2, mulPoint(pk, rho)),
  };
}

File: apps/chain/internal/ocpshuffle/elgamal.go
package ocpshuffle

import "onchainpoker/apps/chain/internal/ocpcrypto"

func elgamalReencrypt(pk ocpcrypto.Point, ct ocpcrypto.ElGamalCiphertext, rho ocpcrypto.Scalar) ocpcrypto.ElGamalCiphertext {
	return ocpcrypto.ElGamalCiphertext{
		C1: ocpcrypto.PointAdd(ct.C1, ocpcrypto.MulBase(rho)),
		C2: ocpcrypto.PointAdd(ct.C2, ocpcrypto.MulPoint(pk, rho)),
	}
}

